pkg/netpol/eval/resources.go

```diff
@@ -203,12 +207,34 @@ func (pe *PolicyEngine) upsertNamespace(ns *corev1.Namespace) error {
 	return nil
 }
 
+// checkLegalOwnerRefNameToPodLabels returns error if there are pod resources with same ownerReferences name but different labels
+func (pe *PolicyEngine) checkLegalOwnerRefNameToPodLabels(ownerRefName string, podLabels map[string]string) error {
+	if ownerRefName == "" { // pod does not have owner references
+		return nil
+	}
+	ownerLabels, ok := pe.ownersToLabelsMap[ownerRefName]
+	if !ok { // add the new ownerReference with pod labels to the map
+		pe.ownersToLabelsMap[ownerRefName] = podLabels
+		return nil
+	}
+	// compare ownerLabels with podLabels
+	if !reflect.DeepEqual(ownerLabels, podLabels) {
+		return errors.New("pods with same ownerReferences but different labels are not supported")
+	}
+	return nil
+}
+
 func (pe *PolicyEngine) upsertWorkload(rs interface{}, kind string) error {
 	pods, err := k8s.PodsFromWorkloadObject(rs, kind)
 	if err != nil {
 		return err
 	}
-	for _, podObj := range pods {
+	for index, podObj := range pods {
+		if index < 1 { // check only once, as all replicas has same ownerReference and same labels
+			if err := pe.checkLegalOwnerRefNameToPodLabels(podObj.Owner.Name, podObj.Labels); err != nil {
```

Comment:
Should be checked only from `upsertPod` and not from `upsertWorkload`.
In the hierarchy of template owners there are cases where controllers add labels for the  generated resources, so that a full match is not necessarily the case. For example, `CronJob` creates pods with added labels `controller-uid`. Another example: `Deployment` creates `ReplicaSet` with labels `pod-template-hash`.